types/src/date/mapping.rs

```diff
@@ -1,23 +1,75 @@
 /*! Mapping for the Elasticsearch `date` type. */
 
 use std::marker::PhantomData;
+use std::borrow::Cow;
 use serde::{Serialize, Serializer};
 use serde::ser::SerializeStruct;
-use super::{DateFormat, Date};
+use super::{DateFormat, FormattedDate, ParsableDate, ParseError, Date, ChronoDateTime};
 use private::field::{DocumentField, FieldMapping, SerializeField};
 use document::FieldType;
 
 /** A field that will be mapped as a `date`. */
-pub trait DateFieldType<M, F>
-    where M: DateMapping<Format = F>,
-          F: DateFormat
+pub trait DateFieldType<M>
 {
 }
 
-impl<T, F, M> FieldType<M, DateFormatWrapper<F>> for T
-    where F: DateFormat,
-          M: DateMapping<Format = F>,
-          T: DateFieldType<M, F> + Serialize
+/** A more general `DateFieldType` that can be parsed and formatted. */
+pub trait DateType where Self: Sized {
```

Comment:
I'm not so sure about the `raw_date` methods. It captures our need to convert to and from a `chrono` date but I don't think that's necessarily valid for callers to do. It's a detail that makes the format methods auto implementable.